source/helpers.js

```diff
@@ -15,3 +36,123 @@ exports.generateTextBar = function (numerator, denominator, barLength) {
 	}
 	return bar;
 }
+
+exports.getNumberEmoji = function (number) {
+	switch (Number(number)) {
+		case 0:
+			return '0️⃣';
+		case 1:
+			return '1️⃣';
+		case 2:
+			return '2️⃣';
+		case 3:
+			return '3️⃣';
+		case 4:
+			return '4️⃣';
+		case 5:
+			return '5️⃣';
+		case 6:
+			return '6️⃣';
+		case 7:
+			return '7️⃣';
+		case 8:
+			return '8️⃣';
+		case 9:
+			return '9️⃣';
+		case 10:
+			return '🔟';
+		default:
+			return '#️⃣';
+	}
+}
+
+/** Recalculates the ranks (standard deviations from mean) and placements (ordinal) for the given participants
+ * @param {Hunter[]} participants
+ * @param {GuildRank[]} ranks
+ * @returns Promise of the message congratulating the hunter reaching first place (or `null` if no change)
+ */
+exports.setRanks = async (participants, ranks) => {
+	let previousFirstPlaceId;
+	let mean = 0;
+	const rankableHunters = [];
+	for (const hunter of participants) {
+		if (hunter.isRankEligible) {
+			if (hunter.seasonPlacement == 1) {
+				previousFirstPlaceId = hunter.userId;
+			}
+			hunter.lastRank = hunter.rank;
+			mean += hunter.seasonXP;
+			rankableHunters.push(hunter);
+		} else {
+			hunter.nextRankXP = 0;
+		}
+	}
+	const n = Math.max(rankableHunters.length, 2);
+	mean /= n;
+	const stdDev = Math.sqrt(rankableHunters.reduce((total, hunter) => total + (hunter.seasonXP - mean) ** 2, 0) / n);
+	for (const hunter of rankableHunters) {
+		let variance = (hunter.seasonXP - mean) / stdDev;
+		ranks.forEach((rank, index) => {
+			if (variance >= rank.varianceThreshold) {
+				hunter.rank = index;
```

Comment:
Still storing this by order index on the hunter seems...volatile? Now that we have a real database with relations, we could store a real reference to the rank in the DB instead.

Reply:
GuildRank uses a composite key of guildId and varianceThreshold. I guess we could decided on a hashing convention?

Reply:
Personally, I'm not looking for a hashing convention. We *should* be able to have `Rank` directly on the `Hunter` object as a foreign key, and then `Hunter` will automagically have both foreign keys (if done right).

Reply:
Sounds fancy!

Reply:
Eh, the hashing function thing is the fancier bit, imo.